packages/@lwc/template-compiler/src/shared/renderer-hooks.ts

```diff
@@ -0,0 +1,109 @@
+ */
+export interface CustomRendererElementConfig {
+    /**
+     * Tag name to use to match an element.
+     */
+    tagName: string;
+    /**
+     * Optional namespace to match an element.
+     */
+    namespace?: string;
+    /**
+     * Specify attributes that need to be matched.
+     * This field is optional.
+     * If undefined or empty, attribute matching is skipped.
+     */
+    attributes?: string[];
+}
+
+/**
+ * Config to specify which elements and directives require a customizable renderer.
+ * An element qualified if it matches the CustomRendererElementConfig OR the directives.
+ */
+export interface CustomRendererConfig {
+    /**
+     * Element matching criteria. Element much satisfy all conditions of the CustomRendererElementConfig
+     */
+    elements: CustomRendererElementConfig[];
+    /**
+     * List of lwc directives that qualify an element. An element must use at least 1
+     * directive to be considered a match
+     */
+    directives: string[];
+}
+
+/**
+ * Mapping of Directive.name to string literals used in template
+ */
+export const LWC_DIRECTIVES: Record<ElementDirective['name'], string> = {
+    Dom: 'lwc:dom',
+    Dynamic: 'lwc:dynamic',
+    InnerHTML: 'lwc:inner-html',
+    Key: 'key',
+};
+
+function checkElement(element: BaseElement, state: State): boolean {
+    let addCustomRenderer = false;
+    const { attributes, directives } = element;
+    if (directives.length) {
+        // If any directives require custom renderer
+        addCustomRenderer = directives.some((dir) => {
+            return state.crDirectives.has(LWC_DIRECTIVES[dir.name]);
+        });
+        if (addCustomRenderer) {
+            // Directives that require custom renderer are not allowed on custom elements
+            // Custom element cannot be allowed to have a custom renderer hook
+            // The renderer is cascaded down from the owner(custom element) to all its child nodes who
+            // do not have a renderer specified.
+            invariant(
+                element.type !== 'Component',
+                TemplateErrors.DIRECTIVE_DISALLOWED_ON_CUSTOM_ELEMENT,
+                [element.name, state.config.customRendererConfig!.directives.join(', ')]
+            );
+        }
+    }
+    const elementConfig = state.crElmToConfigMap[element.name];
+    // If element requires custom renderer
+    if (elementConfig) {
+        const { namespace, attributes: attrConfig } = elementConfig;
+        // if element config has namespace, then namespace has to be a match
+        if (namespace && element.namespace !== namespace) {
+            return false;
+        }
+        // If no attributes are specified, then consider the element requires custom renderer
+        if (
+            attrConfig.size === 0 ||
+            attributes.some((attribute) => attrConfig.has(attribute.name))
+        ) {
+            addCustomRenderer = true;
+        }
+    }
+    return addCustomRenderer;
```

Comment:
The `addCustomRenderer` logic here is a bit confusing to me. I think this relates to the question of how `directives` interacts with `elementConfig`. For instance:

1. What if `addCustomRenderer` is set to `true` on line 63, but then we hit line 84 and return `false`? Should we have returned `true` early instead?
2. On line 91, can we just `return true`?

Reply:
Great catch!